slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -279,7 +279,73 @@ ::grpc::Status SLAMServiceImpl::GetCurrentPointCloudMap(
 ::grpc::Status SLAMServiceImpl::GetPointCloudMapStream(
     ServerContext *context, const GetPointCloudMapStreamRequest *request,
     ServerWriter<GetPointCloudMapStreamResponse> *writer) {
-    return ::grpc::Status(::grpc::StatusCode::UNIMPLEMENTED, "");
+    std::string pointcloud_map;
+    int num_points;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            num_points = GetLatestSampledPointCloudMapString(pointcloud_map);
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_pointcloud_map = pointcloud_map;
+        } else {
+            // We couldn't lock the mutex which means the optimization process
+            // locked it and we need to use the backed up latest map
+            LOG(INFO)
+                << "Optimization is occuring, using cached pointcloud map";
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
+        std::terminate();
+    }
+
+    if (pointcloud_map.empty()) {
+        LOG(ERROR) << "map pointcloud does not have points yet";
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "map pointcloud does not have points yet");
+    }
+
+    std::string pcd_chunk;
+    GetPointCloudMapStreamResponse response;
+
+    // Calculate chunk sizes to ensure no points data is split between chunks
```

Comment:
Why is it important that no points data is split between chunks? Can the receiver handler split points on their end?

Reply:
@tessavitabile My recommendation to not send split chunks is so that the JS UI could (if they choose to) render the points as they come in w/o needing to worry about whether or not a given point is split between chunks or not. I'm not sure what primitives exist on the JS side for handling partial points, so it seemed reasonable to me to have the slam lib be required to send PCD chunks rounded to the nearest point, as doing so would allow every chunk to be self-descriptive once you have the header.

That said if you disagree b/c that requires 3rd party slam libs to do work that is not described in the proto, I'm good to change the design to just send chunks of the max chunk size until the last chunk.

Reply:
I guess my reservations are 

(1) as you said, it requires 3rd party slam libs to do work not described in proto
(2) the chunks will not be full pcds anyway, since we only include the header in the first chunk (and it includes metadata for the whole pcd)

Reply:
Fair enough. Sorry for leading you astray @jeremyhyde-viam , you can remove the requirement to chunk to the nearest point.

Reply:
No worries, reverted changes and removed num_points as it is no longer necessary